File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/flyweight/solution/single_journey_flyweight_factory.cpp
// single_journey_flyweight_factory.cpp

#include <string>
#include "dp_map.h"
#include "single_journey.h"
#include "single_journey_flyweight_factory.h"

single_journey_flyweight_factory::single_journey_flyweight_factory()
{
}

single_journey * single_journey_flyweight_factory::get_flyweight_for(const std::string & a_address)
{
	if (!my_map.contains(a_address))
	{
		my_map.put(a_address, new single_journey(a_address));
	}

	single_journey * result = my_map.get(a_address);

	return result;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/singleton/flyweight/solution/SingleJourneyFlyweightFactory.java
// SingleJourneyFlyweightFactory.java

public class SingleJourneyFlyweightFactory
{
	// Singleton:
	
	private SingleJourneyFlyweightFactory()
	{
		myMap = new DpMap<String, SingleJourney>();
	}
	
	public static SingleJourneyFlyweightFactory getInstance()
	{
		if (theInstance == null)
		{
			theInstance = new SingleJourneyFlyweightFactory();
		}
		
		return theInstance;
	}
	
	private static SingleJourneyFlyweightFactory theInstance = null;
	
	// non-Singleton:
	
	public SingleJourney getFlyweightFor(String aAddress)
	{
		if (!myMap.contains(aAddress))
		{
			myMap.put(aAddress, new SingleJourney(aAddress));
		}

		SingleJourney result = myMap.get(aAddress);

		return result;
	}

	private DpMap<String, SingleJourney> myMap;
}
